githubapp/githubapp.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package githubapp contains a class with methods for any service that needs
+// to interact with GitHub as an app.
+package githubapp
+
+import (
+	"bytes"
+	"context"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/abcxyz/pkg/cache"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+)
+
+// URL used to retrieve access tokens. The pattern must contain a single '%s' which represents where in the url
+// to insert the installation id.
+const defaultGitHubAccessTokenURLPattern = "https://api.github.com/app/installations/%s/access_tokens" //nolint
+
+const cacheKey = "github-app-jwt"
+
+// GitHubAppConfig contains all of the required configuration for operating as a GitHub App.
+// This includes the three major components, the App ID, the Install ID and the Private Key.
+type GitHubAppConfig struct {
+	AppID                 string
+	InstallationID        string
+	PrivateKey            *rsa.PrivateKey
+	accessTokenURLPattern string
+	jwtTokenExpiration    time.Duration
+	jwtCacheDuration      time.Duration
+}
+
+// GitHubAppConfigOption is a function type that applies a mechanism to set optional
+// configuration values.
+type GitHubAppConfigOption func(f *GitHubAppConfig)
```

Comment:
Nit: this name stutters (githubapp.GitHubAppConfig). Maybe name it `Option`?

Reply:
sounds good

Reply:
I renamed GitHubAppConfig to Config, and GitHubAppConfig Option to ConfigOption. I couldn't think of a replacement for GitHubApp but I'm open to it if you have a suggestion.